File: Longest_Valid_Parentheses.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
  public:
    int longestValidParentheses(string s) {
        stack<int> st;
        int maxLength = 0, curLength = 0;

        for (int i = 0; i < s.size(); i++) {
            if (!st.empty() && s[i] == ')' && s[st.top()] == '(')
                st.pop();
            else
                st.push(i); // push the index

            if (!st.empty())
                curLength = i - st.top();
            else
                curLength = i + 1;
            maxLength = maxLength > curLength ? maxLength : curLength;
        }
        return maxLength;
    }
};

File: Longest_Valid_Parentheses.ts
function longestValidParentheses(s: string): number {
    let maxNum = 0
    // the longest string length in previous string
    let numStack: number[] = []
    let stack: string[] = []
    let curNum = 0
    for (const char of s) {
        if (char == '(' || stack.length == 0) {
            stack.push(char)
            if (curNum > 0) {
                numStack.push(curNum)
            }
            numStack.push(0)
            curNum = 0
        } else {
            if (stack[stack.length - 1] == '(') {
                stack.pop()
                numStack.pop()
                curNum += 2
                while (numStack[numStack.length - 1]) {
                    curNum += numStack.pop()!
                }
                maxNum = Math.max(maxNum, curNum)
            } else {
                stack.push(char)
                curNum = 0
            }
        }
    }
    return maxNum
};

let testCases = ["(())", "(())(())", "((()))())", ")(()()))", "", "))", "()(()"]
// let testCases = ["(())(())"]
for (let t of testCases)
    console.log(longestValidParentheses(t))